crates/ra_hir_expand/src/util.rs

```diff
@@ -0,0 +1,114 @@
+//! Internal utility functions.
+//!
+//! The `unquote funciton are mainly copy from enquote crate
+
+use std::error;
+use std::fmt;
+
+#[derive(Debug, PartialEq)]
+pub(crate) enum QuoteError {
+    NotEnoughChars,
+    UnrecognizedQuote,
+    UnexpectedEOF,
+    IllegalChar,
+    UnrecognizedEscape,
+    InvalidUnicode,
+}
+
+impl fmt::Display for QuoteError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self)
+    }
+}
+
+impl error::Error for QuoteError {
+    fn description(&self) -> &str {
+        match self {
+            QuoteError::NotEnoughChars => "not enough chars",
+            QuoteError::UnrecognizedQuote => "unrecognized quote character",
+            QuoteError::UnexpectedEOF => "unexpected eof",
+            QuoteError::IllegalChar => "illegal character",
+            QuoteError::UnrecognizedEscape => "unrecognized escape sequence",
+            QuoteError::InvalidUnicode => "invalid unicode code point",
+        }
+    }
+}
+
+/// Unquotes `s`.
+pub(crate) fn unquote(s: &str) -> Result<String, QuoteError> {
+    if s.chars().count() < 2 {
+        return Err(QuoteError::NotEnoughChars);
+    }
+
+    let quote = s.chars().next().unwrap();
+
+    if quote != '"' && quote != '\'' && quote != '`' {
+        return Err(QuoteError::UnrecognizedQuote);
+    }
+
+    if s.chars().last().unwrap() != quote {
+        return Err(QuoteError::UnexpectedEOF);
+    }
+
+    // removes quote characters
+    // the sanity checks performed above ensure that the quotes will be ASCII and this will not
+    // panic
+    let s = &s[1..s.len() - 1];
+
+    unescape(s, Some(quote))
+}
+
+/// Returns `s` after processing escapes such as `\n` and `\x00`.
+pub(crate) fn unescape(s: &str, illegal: Option<char>) -> Result<String, QuoteError> {
```

Comment:
This already exists as `rustc_lexer::unescape`. We probably should add convenience methods to `ast::StringLiteral` for convenient unescaping

Reply:
actally, i've quickly added in #2280